akka-stream/src/main/scala/akka/stream/impl/fusing/FoldWithin.scala

```diff
@@ -0,0 +1,148 @@
+ * In this case, the output channel needs to be closed if there is no data arriving within certain time to avoid connection timeout.
+ * This custom flow uses custom aggregator to support such use cases.
+ * Upstream inputs are continuously aggregated as they arrive.
+ * The aggregator is terminated based on custom condition, interval gap and total duration using timers.
+ *
+ * @param seed        initiate the aggregated output with first input
+ * @param aggregate   sequentially aggregate input into output
+ * @param emitReady   decide whether the current aggregated output can be emitted
+ * @param maxGap      maximum allowed gap between aggregations, will emit if the interval passed this threshold
+ * @param maxDuration optional total duration for the entire aggregator, requiring a separate timer.
+ * @param harvest     this is invoked as soon as all conditions are met before emitting to next stage, which could take more time.
+ *                    time sensitive operations can be added here, such as closing an output channel
+ * @tparam In
+ * @tparam Agg
+ */
+class FoldWithin[In, Agg, Out](
+                                val seed: In => Agg,
+                                val aggregate: (Agg, In) => Agg,
+                                val emitReady: Agg => Boolean,
+                                val harvest: Agg => Out,
+                                val maxGap: Option[FiniteDuration] = None,
+                                val maxDuration: Option[FiniteDuration] = None
+                              )
+  extends GraphStage[FlowShape[In, Out]] {
+
+  // maxDuration must not be smaller than the maxGap, otherwise maxGap is meaningless
+  for {
+    md <- maxDuration
+    mg <- maxGap
+  } require(md.gteq(mg), s"maxDuration($md) must not be smaller than maxGap($mg)")
+
+  val in: Inlet[In] = Inlet[In](s"${this.getClass.getName}.in")
+  val out: Outlet[Out] = Outlet[Out](s"${this.getClass.getName}.out")
+  override val shape: FlowShape[In, Out] = FlowShape(in, out)
+
+  val maxDurationTimer = "maxDurationTimer"
+  val maxGapTimer = "maxGapTimer"
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+    new TimerGraphStageLogic(shape) with InHandler with OutHandler {
+
+      override def preStart(): Unit = {
+        // although we only need this timer when the aggregator is not empty
+        // and this event do not need to be triggered otherwise,
+        // it's more efficient to set up a upfront timer
+        maxGap.foreach(mg => scheduleWithFixedDelay(maxGapTimer, mg, mg))
+      }
+      override protected def onTimer(timerKey: Any): Unit = state.harvestOnTimer()
+
+      override def onPush(): Unit = {
+        state.aggregate()
+        if (isAvailable(out)) pull(in) // pull only if there is no backpressure from outlet
+      }
+
+      override def onUpstreamFinish(): Unit = {
+        state.harvestOnFlush()
+        completeStage()
```

Comment:
What do you mean "aggregate is not complete"?
Even the condition is not met, it'll harvest whatever left there.
There could be more data pending in upstream, but as long as downstream has no more request, those extra data will not be pulled into the stage. And since upstream is also completed, there is no way to access those data at this point.
Maybe it's helpful to point to the back pressure tests if you think some scenario is not covered.

Reply:
You are right, `harvestAndEmit()` is not conditional, not sure why I thought that. Good 👍

Reply:
Now I remember my thought from yesterday: since this stage does not always grab the element on push, depending on output readiness, the input is in fact used as an additional single element hidden buffer, if there is an element not grabbed when we get here, that should be handed to the aggregate function before calling `harvestAndEmit` so that it isn't lost.

Reply:
That's also wrong though, it always grabs on push, sorry. Disregard my latest comment.

Reply:
I realized most of these tricky cases is due to the previous implementation synchronizes upstream and downstream.
I feel with the more general internal buffering, when/where to grab/pull/emit is more clear.